### PR TITLE
More test generation (take 2)

### PR DESCRIPTION
**Pull request by cpu**
I'm working on test case generation for in-progress CRL support. Before going too far down that road I wanted to try and land the new coverage @ctz started in https://github.com/rustls/webpki/pull/26, integrating some of the feedback from @djc to hopefully make it easier to review. I recommend reading the diff commit-by-commit for maximum approach-ability.

In a perfect world we wouldn't need to use a second language for test generation, but I think Python is well suited to this use case since `cryptography` can do so much of the heavy lifting for us. It would also be a fair bit of work to replace entirely. 

With the above in mind I decided it was prudent to invest in tooling. Beyond bringing in new test coverage this branch:

* Fixes a timezone bug that would generate expired certificates in some cases. 
* Adds a `requirements.txt` describing the dependencies we require.
* Formats the code with [Black](https://pypi.org/project/black/).
* Fixes linting failures flagged by [flake8](https://flake8.pycqa.org/en/latest/).
* Adds static type checking w/ [mypy](https://www.mypy-lang.org/).
* Runs `cargo fmt` automatically after generating new code to save a manual step (or CI failure)
* Adds CI coverage to enforce that we don't regress on formatting, linting, typing or have bit rot in the test case generation tooling or tests.

### CODE REVIEW

File: tests/name_constraints.rs [hunk 1]
```diff
@@ -56,7 +56,7 @@ fn check_cert(
     Ok(())
 }
 
-// DO NOT EDIT BELOW: generated by name_constraints/generate.py
+// DO NOT EDIT BELOW: generated by tests/generate.py
```
Comment (djc) [inline tests/name_constraints.rs:59]: I think this should be in the same commit where the new boundary is used?

File: .github/workflows/testgen.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  linting:
+    runs-on: ubuntu-latest
+    name: Lint
+    steps:
+      - name: Check out source repository
+        uses: actions/checkout@v3
+      - name: Set up Python environment
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.10"
```
Comment (djc) [inline .github/workflows/testgen.yml:17]: Maybe go with 3.11 now?

**Review by djc — COMMENTED** [review]
This is great, thanks for all the small commits.

One thing: in Python scripts I usually wrap module-level runner code into an `if __name__ == '__main__':` block, might be nice to do that for `generate.py`?

File: tests/name_constraints.rs (hunk 1 above)
Comment (cpu) [reply tests/name_constraints.rs:59]: Moved into the re-generation commit.

**Comment by cpu** [comment]
> One thing: in Python scripts I usually wrap module-level runner code into an if __name__ == '__main__': block, might be nice to do that for generate.py?

Sounds good. I did that and also added a shebang. Thanks for the review :-)

**Comment by cpu** [comment]
I have some additional refactoring from writing CRL focused tests on top of WIP work.

@djc Would you prefer I pull that additional refactoring into this branch so we can get it all done with at once, or would you prefer to keep this branch's scope as-is since you've already done a review pass? If you prefer the latter I can make a separate follow-up PR. LMK :-)

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Going to consider this done!

**Pull request merged**